server/src/main/java/org/jboss/as/server/deployment/annotation/CompositeIndexProcessor.java

```diff
@@ -82,16 +84,45 @@ public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitPro
         Map<ModuleIdentifier, CompositeIndex> additionalAnnotationIndexes = new HashMap<ModuleIdentifier, CompositeIndex>();
         final List<ModuleIdentifier> additionalModuleIndexes = deploymentUnit.getAttachmentList(Attachments.ADDITIONAL_ANNOTATION_INDEXES);
         final List<Index> indexes = new ArrayList<Index>();
+
+        Map<ModuleIdentifier, DeploymentUnit> subdeploymentDependencies = buildSubdeploymentDependencyMap(deploymentUnit);
+
         for (final ModuleIdentifier moduleIdentifier : additionalModuleIndexes) {
             AdditionalModuleSpecification additional = additionalModuleSpecificationMap.get(moduleIdentifier);
             if(additional != null) {
+                final List<Index> moduleIndexes = new ArrayList<>();
                 for(ResourceRoot resource : additional.getResourceRoots()) {
                     ResourceRootIndexer.indexResourceRoot(resource);
                     Index indexAttachment = resource.getAttachment(Attachments.ANNOTATION_INDEX);
                     if(indexAttachment != null) {
                         indexes.add(indexAttachment);
+                        moduleIndexes.add(indexAttachment);
+                    }
+                }
+                if (!moduleIndexes.isEmpty()) {
+                    additionalAnnotationIndexes.put(moduleIdentifier, new CompositeIndex(moduleIndexes));
+                }
+            } else if (subdeploymentDependencies.containsKey(moduleIdentifier)) {
+                List<ResourceRoot> resourceRoots = subdeploymentDependencies.get(moduleIdentifier).getAttachment(Attachments.RESOURCE_ROOTS);
+                final List<ResourceRoot> allResourceRoots = new ArrayList<ResourceRoot>();
```

Comment:
The type in the ArrayList can be inferred by the compiler; I would suggest replacing it by new ArrayList<>()